src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+        :class:`Fetcher` instances).
+
+        It need not be an instance of :class:`requests.Session`; one can
+        create any class as long as it implements the :class:`Session`
+        protocol.
+
+    Attributes
+    ----------
+    close_models
+        If true (the default), the cached models will be closed when the
+        fetcher is closed. If set to false, the user must close the models.
+        This is only intended for use by :func:`fetch_model`.
+    close_session
+        If true, the session used for HTTP requests will be closed when the
+        fetcher is closed. It defaults to true if an internally-created session
+        is used, and false if the user provided a session.
+
+    Raises
+    ------
+    :exc:`.ModelError`
+        For any issues with the model itself
+    :exc:`requests.exceptions.RequestException`
+        For any issues at the HTTP level
+    """
+
+    def __init__(self, *, session: Optional[Session] = None) -> None:
+        self._session: Session
+        if session is None:
+            self._session = requests.Session()
+            # TODO: requests_file is convenient, but it would be more efficient to
+            # open the file directly with h5py rather than sucking it into a
+            # BytesIO.
+            self._session.mount('file://', requests_file.FileAdapter())
+            self.close_session = True
+        else:
+            self._session = session
+            self.close_session = False
+        self._alias_cache: Dict[str, str] = {}
+        self._model_cache: Dict[str, models.Model] = {}
+        self.close_models = True
+
+    @property
+    def session(self) -> Session:
+        return self._session
+
+    def close(self) -> None:
+        """Release the resources associated with the fetcher.
+
+        See also
+        --------
+        :attr:`close_session`, :attr:`close_models`
+        """
+        if self.close_session:
+            self._session.close()
+        self._alias_cache.clear()
+        if self.close_models:
+            for model in self._model_cache.values():
+                model.close()
+        self._model_cache.clear()
+
+    def resolve(self, url: str) -> str:
+        """Find the canonical URL, following aliases."""
+        original_url = url
+        aliases = 0
+        parts = urllib.parse.urlparse(url)
+        while parts.path.endswith('.alias'):
+            aliases += 1
+            if aliases > MAX_ALIASES:
+                raise models.TooManyAliasesError.with_urls(
+                    f'Reached limit of {MAX_ALIASES} levels of aliases',
+                    url=url, original_url=original_url)
+            if url in self._alias_cache:
+                new_url = self._alias_cache[url]
+            else:
+                with self._session.get(url) as resp:
+                    resp.raise_for_status()
+                    rel_path = resp.text.rstrip()
+                    new_url = urllib.parse.urljoin(resp.url, rel_path)
```

Comment:
Maybe check for loops when new_url == url

Reply:
I'm planning to change this to return the entire path, which I need for something else; once I've done that it becomes easy to check for any sort of loop.

Reply:
Updated to return the resolution path, and added explicit loop detection.